activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1195,6 +1251,37 @@ def build_select(arel)
         end
       end
 
+      def build_with(arel)
+        return if with_values.empty?
+
+        is_recursive = with_values.delete(:recursive)
+        with_statements = with_values.map do |with_value|
+          case with_value
+          when String then Arel::Nodes::SqlLiteral.new(with_value)
+          when Arel::Nodes::As then with_value
+          when Array
+            raise ArgumentError, "Unsupported argument type: #{with_value} #{with_value.class}" unless with_value.map(&:class).uniq == [Arel::Nodes::As]
+
+            with_value
+          when Hash then
+            with_value.map do |name, expression|
+              table = Arel::Table.new(name)
+              expression = expression.arel if expression.is_a?(ActiveRecord::Relation)
```

Comment:
Any chance to keep logic in case/when here and support also `Arel::SelectManager`?

https://github.com/rails/rails/pull/37973/files#diff-bf6dd6226db3aab589916f09236881c7R1205-R1215

Reply:
`Arel::SelectManager` is supported. It is just passed as is. No need to wrap it with `Arel::Nodes::SqlLiteral`

Reply:
Ahh, ok. What do you think to keep case/when logic instead of `if`s here?

Reply:
Ok, I'll check that when I'm back at my laptop later today.

My idea was that `expression` is either `Arel::SelectManager` or `Arel::Nodes:SqlLiteral` in the end. `ActiveRecord::Relation` is transformed to select manager with `arel` method and `String` into sql literal. I could revisit it to use case/when and even add unsupported argument error in else block then.

Reply:
It just looks more readable IMHO. If unsupported expression is passed to `Arel::Nodes::As`, it will raise `TypeError` later during visiting that node. It will probably make sense to just allow supported in when/case and raise on unsupported to not let bubble up this low-level exception up to users.

Reply:
> It just looks more readable IMHO. 

I wanted to avoid nested case/when but I don't have strong opinion about this and agree with your proposal.

> If unsupported expression is passed to `Arel::Nodes::As`, it will raise `TypeError` later during visiting that node. It will probably make sense to just allow supported in when/case and raise on unsupported to not let bubble up this low-level exception up to users.

Yes, that was my idea also to show some meaningful error to the user instead of low-level one.

Reply:
:handshake:

Reply:
I changed the code to use case/when and you were right, looks better now.
Only downside is that you need to know all Arel classes that can be passed (`Arel::SelectManager`, `Arel::Nodes::Union`, ...).